File: code/src-tauri/src/models/worktree.rs
use serde::{Deserialize, Serialize};

/// Git Worktree 状态
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum WorktreeStatus {
    #[serde(rename = "clean")]
    Clean,
    #[serde(rename = "dirty")]
    Dirty,
    #[serde(rename = "unpushed")]
    Unpushed,
    #[serde(rename = "conflicted")]
    Conflicted,
    #[serde(rename = "detached")]
    Detached,
    #[serde(rename = "unknown")]
    Unknown,
}

/// 最后提交信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LastCommit {
    /// 提交 hash (短)
    pub hash: String,
    /// 提交消息 (第一行)
    pub message: String,
    /// 作者
    pub author: String,
    /// 相对时间
    pub relative_time: String,
}

/// Worktree 信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Worktree {
    /// 唯一标识符
    pub id: String,
    /// 名称
    pub name: String,
    /// 分支名
    pub branch: String,
    /// 文件路径
    pub path: String,
    /// 状态
    pub status: WorktreeStatus,
    /// 最后提交信息
    pub last_commit: LastCommit,
    /// 最后活跃时间
    pub last_active_at: Option<String>,
    /// 是否为主 worktree
    pub is_main: bool,
    /// 关联的远程仓库
    pub remote: Option<String>,
}

/// 创建 Worktree 参数
#[derive(Debug, Deserialize)]
pub struct CreateWorktreeParams {
    pub name: String,
    pub base_branch: String,
    pub new_branch: Option<String>,
    pub custom_path: Option<String>,
}

/// Worktree 操作结果
#[derive(Debug, Serialize)]
pub struct WorktreeResult {
    pub success: bool,
    pub message: String,
    pub worktree: Option<Worktree>,
}

/// Worktree 列表响应
#[derive(Debug, Serialize)]
pub struct WorktreeListResponse {
    pub worktrees: Vec<Worktree>,
    pub repo_path: String,
    pub is_valid_repo: bool,
}

/// 分支信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Branch {
    /// 分支名
    pub name: String,
    /// 是否为当前分支
    pub is_current: bool,
}

/// 分支列表响应
#[derive(Debug, Serialize)]
pub struct BranchListResponse {
    pub branches: Vec<Branch>,
    pub current_branch: String,
}

/// Diff 统计信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiffStats {
    /// 文件路径
    pub path: String,
    /// 新增行数
    pub additions: usize,
    /// 删除行数
    pub deletions: usize,
    /// 状态 (added, modified, deleted, renamed)
    pub status: String,
}

/// Diff 响应
#[derive(Debug, Serialize)]
pub struct DiffResponse {
    /// 源分支
    pub source_branch: String,
    /// 目标分支
    pub target_branch: String,
    /// 文件变更统计
    pub files: Vec<DiffStats>,
    /// 总新增行数
    pub total_additions: usize,
    /// 总删除行数
    pub total_deletions: usize,
    /// 变更文件数
    pub files_changed: usize,
}

/// Diff 行
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiffLine {
    /// 行类型: "context" | "addition" | "deletion" | "header"
    pub line_type: String,
    /// 旧文件行号 (删除行和上下文行)
    pub old_line: Option<usize>,
    /// 新文件行号 (新增行和上下文行)
    pub new_line: Option<usize>,
    /// 行内容
    pub content: String,
}

/// Diff Hunk (代码块)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiffHunk {
    /// 旧文件起始行
    pub old_start: usize,
    /// 旧文件行数
    pub old_lines: usize,
    /// 新文件起始行
    pub new_start: usize,
    /// 新文件行数
    pub new_lines: usize,
    /// 行内容
    pub lines: Vec<DiffLine>,
}

/// 文件详细 Diff
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileDiff {
    /// 文件路径
    pub path: String,
    /// 旧文件路径 (重命名时)
    pub old_path: Option<String>,
    /// 状态
    pub status: String,
    /// Hunks
    pub hunks: Vec<DiffHunk>,
    /// 新增行数
    pub additions: usize,
    /// 删除行数
    pub deletions: usize,
}

/// 详细 Diff 响应
#[derive(Debug, Serialize)]
pub struct DetailedDiffResponse {
    /// 源分支
    pub source_branch: String,
    /// 目标分支
    pub target_branch: String,
    /// 文件列表
    pub files: Vec<FileDiff>,
    /// 总新增行数
    pub total_additions: usize,
    /// 总删除行数
    pub total_deletions: usize,
}

/// 仓库信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RepositoryInfo {
    /// 仓库 ID (路径)
    pub id: String,
    /// 仓库名称
    pub name: String,
    /// 路径
    pub path: String,
    /// 当前分支
    pub current_branch: String,
    /// Worktree 数量
    pub worktree_count: usize,
    /// 最后活跃时间
    pub last_active: Option<String>,
}

/// 切换分支结果
#[derive(Debug, Serialize)]
pub struct SwitchBranchResult {
    pub success: bool,
    pub message: String,
}

/// 批量删除结果
#[derive(Debug, Serialize)]
pub struct BatchDeleteResult {
    pub success_count: usize,
    pub failed_count: usize,
    pub results: Vec<WorktreeResult>,
}

/// Worktree 提示信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorktreeHint {
    /// Worktree ID
    pub worktree_id: String,
    /// 分支名
    pub branch: String,
    /// 提示类型: "merged" | "stale"
    pub hint_type: String,
    /// 提示消息
    pub message: String,
    /// 是否已合并
    pub is_merged: bool,
    /// 最后活跃天数
    pub inactive_days: Option<i64>,
}

File: code/src/types/worktree.ts
/**
 * Git Worktree 状态枚举
 */
export enum WorktreeStatus {
  /** 干净状态，无未提交更改 */
  Clean = 'clean',
  /** 有未提交的更改 */
  Dirty = 'dirty',
  /** 有未推送的提交 */
  Unpushed = 'unpushed',
  /** 有冲突 */
  Conflicted = 'conflicted',
  /** Detached HEAD 状态 */
  Detached = 'detached',
  /** 未知状态 */
  Unknown = 'unknown',
}

/**
 * 远程同步状态
 */
export interface SyncStatus {
  /** 领先远程的提交数 */
  ahead: number
  /** 落后远程的提交数 */
  behind: number
  /** 是否有远程分支 */
  hasRemote: boolean
}

/**
 * 最后提交信息
 */
export interface LastCommit {
  /** 提交 hash (短) */
  hash: string
  /** 提交消息 (第一行) */
  message: string
  /** 作者 */
  author: string
  /** 相对时间 */
  relativeTime: string
}

/**
 * Worktree 信息
 */
export interface Worktree {
  /** Worktree 唯一标识符 */
  id: string
  /** Worktree 名称 */
  name: string
  /** 所在分支名 */
  branch: string
  /** 文件系统路径 */
  path: string
  /** 当前状态 */
  status: WorktreeStatus
  /** 最后提交信息 */
  lastCommit: LastCommit
  /** 最后活跃时间 */
  lastActiveAt: string | null
  /** 是否为主 Worktree */
  isMain: boolean
  /** 关联的远程仓库名 */
  remote?: string
  /** 额外元数据 */
  metadata?: Record<string, unknown>
  /** 远程同步状态 */
  syncStatus: SyncStatus
}

/**
 * 创建 Worktree 请求参数
 */
export interface CreateWorktreeParams {
  /** Worktree 名称 */
  name: string
  /** 基于的分支名 */
  baseBranch: string
  /** 新分支名（可选，不提供则自动生成） */
  newBranch?: string
  /** 自定义路径（可选） */
  customPath?: string
}

/**
 * Worktree 操作结果
 */
export interface WorktreeResult {
  /** 是否成功 */
  success: boolean
  /** 结果消息 */
  message: string
  /** 操作后的 Worktree 信息（可选） */
  worktree?: Worktree
}

/**
 * Worktree 列表响应
 */
export interface WorktreeListResponse {
  /** Worktree 列表 */
  worktrees: Worktree[]
  /** 当前仓库路径 */
  repoPath: string
  /** 是否是有效的 Git 仓库 */
  isValidRepo: boolean
}

/**
 * 分支信息
 */
export interface Branch {
  /** 分支名 */
  name: string
  /** 是否为当前分支 */
  isCurrent: boolean
  /** 最后提交信息 */
  lastCommit?: string
  /** 最后提交时间 */
  lastCommitDate?: string
  /** 关联的远程分支 */
  upstream?: string
}

/**
 * 分支列表响应
 */
export interface BranchListResponse {
  /** 分支列表 */
  branches: Branch[]
  /** 当前分支名 */
  currentBranch: string
}

/**
 * Diff 统计信息
 */
export interface DiffStats {
  /** 文件路径 */
  path: string
  /** 新增行数 */
  additions: number
  /** 删除行数 */
  deletions: number
  /** 状态 (added, modified, deleted, renamed) */
  status: string
}

/**
 * Diff 响应
 */
export interface DiffResponse {
  /** 源分支 */
  sourceBranch: string
  /** 目标分支 */
  targetBranch: string
  /** 文件变更统计 */
  files: DiffStats[]
  /** 总新增行数 */
  totalAdditions: number
  /** 总删除行数 */
  totalDeletions: number
  /** 变更文件数 */
  filesChanged: number
}

/**
 * Diff 行
 */
export interface DiffLine {
  /** 行类型: "context" | "addition" | "deletion" */
  lineType: string
  /** 旧文件行号 */
  oldLine: number | null
  /** 新文件行号 */
  newLine: number | null
  /** 行内容 */
  content: string
}

/**
 * Diff Hunk (代码块)
 */
export interface DiffHunk {
  /** 旧文件起始行 */
  oldStart: number
  /** 旧文件行数 */
  oldLines: number
  /** 新文件起始行 */
  newStart: number
  /** 新文件行数 */
  newLines: number
  /** 行内容 */
  lines: DiffLine[]
}

/**
 * 文件详细 Diff
 */
export interface FileDiff {
  /** 文件路径 */
  path: string
  /** 旧文件路径 (重命名时) */
  oldPath: string | null
  /** 状态 */
  status: string
  /** Hunks */
  hunks: DiffHunk[]
  /** 新增行数 */
  additions: number
  /** 删除行数 */
  deletions: number
}

/**
 * 详细 Diff 响应
 */
export interface DetailedDiffResponse {
  /** 源分支 */
  sourceBranch: string
  /** 目标分支 */
  targetBranch: string
  /** 文件列表 */
  files: FileDiff[]
  /** 总新增行数 */
  totalAdditions: number
  /** 总删除行数 */
  totalDeletions: number
}

/**
 * 仓库信息
 */
export interface RepositoryInfo {
  /** 仓库 ID (路径) */
  id: string
  /** 仓库名称 */
  name: string
  /** 路径 */
  path: string
  /** 当前分支 */
  currentBranch: string
  /** Worktree 数量 */
  worktreeCount: number
  /** 最后活跃时间 */
  lastActive: string | null
  /** 路径是否有效 */
  isPathValid?: boolean
}

/**
 * 切换分支结果
 */
export interface SwitchBranchResult {
  success: boolean
  message: string
}

/**
 * 批量删除结果
 */
export interface BatchDeleteResult {
  successCount: number
  failedCount: number
  results: WorktreeResult[]
}

/**
 * Worktree 提示信息
 */
export interface WorktreeHint {
  /** Worktree ID */
  worktreeId: string
  /** 分支名 */
  branch: string
  /** 提示类型: "merged" | "stale" */
  hintType: string
  /** 提示消息 */
  message: string
  /** 是否已合并 */
  isMerged: boolean
  /** 最后活跃天数 */
  inactiveDays: number | null
}